File: BalancedBinaryTree/BalBinTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
  if (!root) return true;
  let leftTree = balanced(root.left);
  let rightTree = balanced(root.right);
  if (
    Math.abs(rightTree - leftTree) <= 1 &&
    isBalanced(root.left) === true &&
    isBalanced(root.right) === true
  ) {
    return true;
  }
  return false;
};

function balanced(root) {
  if (root === null) return 0;
  return Math.max(balanced(root.left), balanced(root.right)) + 1;
}

File: BalancedBinaryTree/BalBinTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def balanced(root):
            if root == None:
                return True
            leftTree = balanced(root.left)
            if leftTree == - 1:
                return -1
            rightTree = balanced(root.right)
            if rightTree == - 1:
                return -1
            if abs(leftTree - rightTree) > 1:
                return -1
            return max(leftTree, rightTree) + 1
        return balanced(root) != -1